lib/segment/src/index/vector_index_base.rs

```diff
@@ -58,8 +57,7 @@ pub trait VectorIndex {
 #[derive(Debug)]
 pub enum VectorIndexEnum {
     Plain(PlainIndex),
-    HnswRam(HNSWIndex<GraphLinksRam>),
-    HnswMmap(HNSWIndex<GraphLinksMmap>),
+    Hnsw(HNSWIndex),
```

Comment:
Honestly, it looks inconsistent to me. Each variant has `Ram`/`Mmap`/`Compressed` versions and only HNSW index is an exception with internal definitions. Like quantization and vector storage, we do the same: expand all possible options in a single enum. Future developers may be confused and think that hnsw does not have ram/mmap versions. I won't say it's a blocker, but let's discuss it. Probably we can inline all `HNSWIndex` variants here

Reply:
TBH I don't see an issue here. Sparse is already a different beast than HNSW/dense, and now they are different just a bit more. Whether an enum is expanded here I see an implementation detail of each index.

I did two things here:
- Removed a generic parameter from `HNSWIndex`. Pros: lesser compiler times and binary sizes.
- Removed an enum variant from `VectorIndexEnum` as it duplicates an internal runtime flag of `HNSWIndex`. Among other things, this change makes an [Illegal State Unrepresentable](https://corrode.dev/blog/illegal-state/), i.e. we can't have `on_disk: true` with `HnswRam` by accident.

Reply:
@timvisee needs additional opinion. After a personal discussion with @xzfc , we don't have a strong vision in the same direction

Reply:
My personal unsolicited take: @xzfc makes good arguments, for me it makes sense to go with this new design, which partitions vector indices by nature, then by implementation. 

Regarding consistency, I would argue that all the other `Sparse*` variants should undergo a similar refactoring (joining them into a separate enum)

Reply:
I agree that @xzfc proposal is good. But I believe we won't do the same for all variants and all enums with old design

Reply:
Anyway thanks @coszio! I find your comment as "looks good", resolve this thread

Reply:
Thanks for bringing this to our attention Ivan, it's good to discuss changes with this significance.

I agree that it would be nice to have the same architecture in multiple places, but in my opinion xzfc gives two compelling arguments. Based on that I also vote to keep the design in this PR.